translator/trace/otlp/otlp_to_oc.go

```diff
@@ -0,0 +1,303 @@
+		}
+	}
+
+	return &octrace.Span{
+		TraceId:      span.TraceId,
+		SpanId:       span.SpanId,
+		Tracestate:   tracestateToOC(span.Tracestate),
+		ParentSpanId: span.ParentSpanId,
+		Name: &octrace.TruncatableString{
+			Value: span.Name,
+		},
+		Kind:           kindToOC(span.Kind),
+		StartTime:      unixnanoToTimestamp(span.StartTimeUnixnano),
+		EndTime:        unixnanoToTimestamp(span.EndTimeUnixnano),
+		Attributes:     attributesToOCSpanAttributes(span.Attributes, span.DroppedAttributesCount),
+		TimeEvents:     eventsToOC(span.Events, span.DroppedEventsCount),
+		Links:          linksToOC(span.Links, span.DroppedLinksCount),
+		Status:         statusToOC(span.Status),
+		ChildSpanCount: childSpanCount,
+	}
+}
+
+func linksToOC(links []*tracepb.Span_Link, droppedCount uint32) *octrace.Span_Links {
+	if len(links) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_Links{
+		Link:              linksToOCLinks(links),
+		DroppedLinksCount: int32(droppedCount),
+	}
+}
+
+func linksToOCLinks(links []*tracepb.Span_Link) []*octrace.Span_Link {
+	if len(links) == 0 {
+		return nil
+	}
+
+	ocLinks := make([]*octrace.Span_Link, 0, len(links))
+	for _, link := range links {
+		ocLink := &octrace.Span_Link{
+			TraceId:    link.TraceId,
+			SpanId:     link.SpanId,
+			Tracestate: tracestateToOC(link.Tracestate),
+			Attributes: attributesToOCSpanAttributes(link.Attributes, link.DroppedAttributesCount),
+		}
+		ocLinks = append(ocLinks, ocLink)
+	}
+	return ocLinks
+}
+
+func eventsToOC(events []*tracepb.Span_Event, droppedCount uint32) *octrace.Span_TimeEvents {
+	if len(events) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_TimeEvents{
+		TimeEvent:                 eventsToOCEvents(events),
+		DroppedMessageEventsCount: int32(droppedCount),
+	}
+}
+
+func eventsToOCEvents(events []*tracepb.Span_Event) []*octrace.Span_TimeEvent {
+	if len(events) == 0 {
+		return nil
+	}
+
+	ocEvents := make([]*octrace.Span_TimeEvent, 0, len(events))
+	for _, event := range events {
+		ocEvent := &octrace.Span_TimeEvent{
+			Time: unixnanoToTimestamp(event.TimeUnixnano),
+			Value: &octrace.Span_TimeEvent_Annotation_{
+				Annotation: &octrace.Span_TimeEvent_Annotation{
+					Description: &octrace.TruncatableString{
+						Value: event.Name,
+					},
+					Attributes: attributesToOCSpanAttributes(event.Attributes, event.DroppedAttributesCount),
+				},
+			},
+		}
+		ocEvents = append(ocEvents, ocEvent)
+	}
+	return ocEvents
+}
+
+func statusToOC(status *tracepb.Status) *octrace.Status {
+	if status == nil {
+		return nil
+	}
+	return &octrace.Status{
+		Code:    int32(status.Code),
+		Message: status.Message,
+	}
+}
+
+func resourceToOC(resource *resourcepb.Resource) (*occommon.Node, *ocresource.Resource) {
+	if resource == nil {
+		return nil, nil
+	}
+
+	return &occommon.Node{
+			Attributes: attributesToOC(resource.Attributes),
+		},
+		&ocresource.Resource{
+			Labels: attributesToOC(resource.Attributes),
+		}
```

Comment:
This will create duplicate sets of labels in Node and Resource. We need to put specially named node attributes in the Node, special resource attributes in the Resource and the rest should go to the Resource. See `ocNodeResourceToOtlp` for special attributes. This translation should be the mirror of it as much as possible.